src/main/java/org/xwiki/git/GitManager.java

```diff
@@ -68,6 +68,25 @@ default Repository getRepository(String repositoryURI, String localDirectoryName
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * Clone a private Git repository as Bare using the credentials provided by user and store it locally in the
+     * XWiki Permanent directory. If the repository is already cloned, no action is done.
+     *
+     * @param repositoryURI the URI to the Git repository to clone (eg "git://github.com/xwiki/xwiki-commons.git")
+     * @param localDirectoryName the name of the directory where the Git repository will be cloned (this directory is
+     *        relative to the permanent directory
+     * @param username the username of the Git user
```

Comment:
not precise enough, doesn't tell that you can pass null. I think it could be best to use Optional.

Reply:
Same for the previous commit/PR, we might want to use Optional.

Reply:
Actually I think it could be best to pass a CredentialsProvider object instead. Offers more possibilities.

Reply:
ofc the script service API would still use strings but I'm referring here to the Java API.

Reply:
For the credentials and bare use case, I'm actually wondering if we shouldn't allow passing a CloneCommand object for complex use cases. This opens up all possibilities of configuring it the way the caller needs it.

Reply:
CredentialsProvider seems like a better option, I am doing the changes.
About complex use cases, shouldn't the caller use his own implementation since this is only about getting the repository and saving it locally?
We can pass CloneCommand object though, but I believe it's better to make a new method for it. WDYT?

Reply:
For the script API sure. For the java API I'm less sure and I think I'd go with passing a CloneCommand. If you look at our implementation you'll find that the clonecommand part is only these few lines and they don't bring much value: https://github.com/xwiki-contrib/api-git/blob/a6e7cf8d40595fed34643029eefac1162d11a757/src/main/java/org/xwiki/git/internal/DefaultGitManager.java#L94-L99 

The value is the rest: the location and the git init. The CloneCommand is just configuration and is the same as passing parameters except that you're nowhere need being able to specify all your can set on a clone command (and what about when jgit adds support for a new config option in CloneCommand)?

Reply:
> and we'll get to 7-8 in no time...

setFs(), setRemote(), setBranch(), setCloneAllBranches(), setMirror(), setCloneSubmodules(), setBranchesToClone(), setTagOption(), setTagOption(), setNoCheckout(), etc

Reply:
True, I now replaced them by CloneCommand.